pkg/operator/ceph/cluster/osd/spec.go

```diff
@@ -74,12 +75,32 @@ set -o nounset # fail if variables are unset
 set -o xtrace
 
 OSD_ID="$ROOK_OSD_ID"
+CEPH_FSID=%s
 OSD_UUID=%s
 OSD_STORE_FLAG="%s"
 OSD_DATA_DIR=/var/lib/ceph/osd/ceph-"$OSD_ID"
 CV_MODE=%s
 DEVICE="$%s"
 
+# "ceph.conf" must have the "fsid" global configuration after the following Ceph's PR is merged.
+# https://github.com/ceph/ceph/commit/25655e5a8829e001adf467511a6bde8142b0a575
+cp --no-preserve=mode /etc/temp-ceph/ceph.conf /etc/ceph/ceph.conf
+python3 -c "
+import configparser
+
+config = configparser.ConfigParser()
+config.read('/etc/ceph/ceph.conf')
+
+if not config.has_section('global'):
+    config['global'] = {}
+
+if not config.has_option('global','fsid'):
+    config['global']['fsid'] = '$CEPH_FSID'
```

Comment:
Until now we have always specified ceph config with command line parameters. Is there no way to specify the fsid as a command line parameter or why is the fsid treated different? Then we don't need to hack the custom ceph.conf overrides.

Reply:
So hopefully we can just add another env var to the init container, or add a command line var to the activate command.

Reply:
IIRC, options like `cephx`, `fsid`, `mon_host` essentially information used to connect cannot be used with the mon store, at least they would require a restart of the daemons to take effect. I believe fsid is one of them. Also, ceph-volume does not go and read the mon store, it only reads the `ceph.conf`.

Reply:
@guits Do you know how we can pass the cluster fsid to `ceph-volume lvm activate` other than in the ceph.conf?

Reply:
> ```
> $ kubectl -n rook-ceph get pod -lapp=rook-ceph-osd
> NAME                               READY   STATUS                  RESTARTS      AGE
> rook-ceph-osd-0-6695b48fd6-6mft7   0/1     Init:CrashLoopBackOff   4 (31s ago)   118s
> $ kubectl -n rook-ceph logs rook-ceph-osd-0-6695b48fd6-6mft7 -c activate
> + ceph-volume lvm activate --fsid=458615f8-843d-4d4f-93cc-76ed42d0e6f7 --no-systemd --bluestore 0 297a266f-f455-48d7-8906-374e97a2cd3f
> usage: ceph-volume lvm activate [-h] [--auto-detect-objectstore] [--bluestore]
>                                 [--filestore] [--all] [--no-systemd]
>                                 [ID] [FSID]
> ceph-volume lvm activate: error: unrecognized arguments: --fsid=458615f8-843d-4d4f-93cc-76ed42d0e6f7
> ```
> 

@satoru-takeuchi @travisn how was prepared this OSD? (what was the c-v command run?)

Reply:
@guits 

> how was prepared this OSD? (what was the c-v command run?)

This OSD was initialized by the following command.

```
ceph-volume --log-path /tmp/ceph-log lvm batch --prepare --bluestore --yes --dmcrypt --osds-per-device 1 /dev/sdb --crush-device-class hdd --report
```

Reply:
Something is unclear to me, I'm curious how this OSD could be prepared without a `ceph.conf` that contains the `fsid` parameter.

```
[root@cluster-fsid-1 /]# ceph-volume lvm batch --prepare --bluestore --yes --dmcrypt --osds-per-device 1 /dev/sda
--> DEPRECATION NOTICE
--> You are using the legacy automatic disk sorting behavior
--> The Pacific release will change the default to --no-auto
--> passed data devices: 1 physical, 0 LVM
--> relative data size: 1.0
Running command: /usr/bin/ceph-authtool --gen-print-key
Running command: /usr/bin/ceph-authtool --gen-print-key
-->  NoOptionError: No option 'fsid' in section: 'global'
[root@cluster-fsid-1 /]# ceph --version
ceph version 16.2.9 (4c3647a322c0ff5a1dd2344e039859dcbd28c830) pacific (stable)
[root@cluster-fsid-1 /]# cat /etc/ceph/ceph.conf
# minimal ceph.conf for c90ee372-e65c-11ec-82c7-525400bf0511
[global]
#       fsid = c90ee372-e65c-11ec-82c7-525400bf0511
        mon_host = [v2:10.10.10.11:3300/0,v1:10.10.10.11:6789/0]
[root@cluster-fsid-1 /]#
```

Reply:
@guits Yes, we have seen that the `fsid` is required in the ceph.conf or else the encrypted osd provisioning will fail as you see. My question is if there is any env var or parameter for `ceph-volume` for the fsid, instead of requiring it in the ceph.conf? Until now Rook has always been able to pass required settings with env vars or parameters, so Rook hasn't needed to even use ceph.conf in many scenarios.

Reply:
From offline conversation, @guits found a fix for this in ceph-volume so the fsid won't be necessary during activate for the encrypted OSDs. But we will need to go ahead and merge this for now to unblock the encrypted OSDs, and later we can revert after the fix is in a new ceph release and we no longer need to support these releases (it will be a while). 
@satoru-takeuchi How about adding a comment to the code that we can remove this after we remove quincy support?

Reply:
found out the root cause. it's a regression introduced by https://github.com/ceph/ceph/pull/44765.
As @travisn mentioned I'm working on a patch in ceph-volume in order to address this issue.

Reply:
https://github.com/ceph/ceph/pull/46588 should address this issue

Reply:
@guits Thank you very much!